Extension improvements (kube version check + annotations persistence)

aalves08 opened the pull request:
Fixes #9132 

- prevent extension load if `local` cluster kube version is not compatible with `uiPlugin` kube version
- improve kube version incompatibility check mechanism for extensions (it was flawed before because it assumed we had the versions available to cross-check it, but if the kube version is incompatible then the extension operator doesn't even load it from the backend, making it non-existent for the cross-check)
- add banner to extensions screen when there's a kube version incompatibility with any of the installed extensions
- add extension chart version annotations to the `uiPlugin` resource on extension installation

**To test**
 - install `elemental` extension (latest version)
 - check that the `uiPlugin` resource created for `elemental` has the annotations from the chart version under `spec.plugin.metadata`
 - quick test for new kube version incompatibility banner: edit https://github.com/rancher/dashboard/blob/master/shell/pages/c/_cluster/uiplugins/index.vue#L127 after install to `v1.29.2` so that the current installed version of `elemental` becomes "incompatible" with the current kube version
 - quick test for not loading extension: edit  `currKubeVersion` variable in `plugin.js` file to `v1.29.12+k3s1`so that the `local` cluster kube version is greater than `v1.28` (`elemental` default)

**Screenshots**
![Screenshot 2023-09-26 at 18 24 31](https://github.com/rancher/dashboard/assets/97888974/c96b3837-62f8-4278-82a8-c840691ce91c)

![Screenshot 2023-09-01 at 13 58 01](https://github.com/rancher/dashboard/assets/97888974/dbf895fa-9608-4f80-9a6e-3ea713c73d3c)

aalves08 commented:
@richard-cox , I've fixed the e2e issue. Do you think we can still move forward with this feature for `v2.8.0`?

richard-cox commented:
This is still a 2.8.0 thing. I'd like to give a quick review

richard-cox reviewed (changes requested):
I didn't get to test functional stuff because we might have some issues, specifically fetching the local cluster to get it's kube version
- user may not have permissions to see the local cluster
- hopefully we should be able to load plugins on the log in screen in 2.8.0. at this point fetching the local cluster is impossible

Couple of options
- we ask that the local kube version be added to the `rancherversion` request (possible security issue?)
- we ask the response to `http:ui-plugin-operator:80/proxy/index.json` to contain info on whether they're valid or not

aalves08 commented on shell/assets/translations/en-us.yaml:
```diff
@@ -4167,6 +4168,8 @@ plugins:
     title: Add Extensions repositories
     prompt: You can install multiple Rancher extension repositories to increase your extensions catalog
+  kubeIncompatibility:
+    banner: There are some installed extensions that may not work properly because they are incompatible with the current Kubernetes version of your local cluster.
```
updated

richard-cox replied:
there was a typo in my suggestion `property` --> `properly`

aalves08 replied:
fixed

aalves08 commented on shell/pages/c/_cluster/uiplugins/index.vue:
```diff
@@ -449,6 +455,27 @@ export default {
       let changes = 0;
 
+      // An install occured and a plugin was added, let add all annotations to the UIPlugin resource (useful for other checks)
+      if (neu?.length === 1 && !installed.find((x) => neu[0].id === x.id) &&
```
updated

aalves08 commented on shell/plugins/plugin.js:
```diff
@@ -29,41 +29,57 @@ export default async function(context) {
+    const reqs = await allHashSettled({
+      // fetch clusters to get local cluster kube version
+      clusters: context.store.dispatch('management/request', {
+        url:                  `v3/clusters`,
```
Cool. Thanks. It's only the `local` cluster needed. Will adjust accordingly

aalves08 commented on shell/plugins/plugin.js:
```diff
@@ -29,41 +29,57 @@ export default async function(context) {
+    const reqs = await allHashSettled({
+      // fetch clusters to get local cluster kube version
+      clusters: context.store.dispatch('management/request', {
+        url:                  `v3/clusters`,
```
True... I'll have to examine the whole logic to check what is the contingency plan for this 🤔 Will present my findings or a solution

aalves08 replied:
Well, I've looked into this and a user needs to be a `local` `Cluster Owner` in order to manage extensions. In that case, we got permission to get the kube version.
Worst case scenario is a standard user that doesn't have access to local cluster will **still** load extensions like he did before and they won't benefit from this update. Not ideal but no harm here (for now... we would need to create a follow-up issue).

Logic implemented covers the scenario where we it doesn't have access to local cluster.

If you are okay with this @richard-cox , we can resolve this conversation. Will leave it up to you.

richard-cox replied:
I'm not sure this is the best path forward. Admins will then not see extensions that their users can

aalves08 commented on shell/pages/c/_cluster/uiplugins/index.vue:
```diff
@@ -449,6 +455,27 @@ export default {
       let changes = 0;
 
+      // An install occured and a plugin was added, let add all annotations to the UIPlugin resource (useful for other checks)
+      if (neu?.length === 1 && !installed.find((x) => neu[0].id === x.id) &&
```
> * `neu?.length === 1` - this only works if there's a single, non-built in plugin. if there's more than one it will never fire?
> * what happens with `latestInstalledPlugin` if two are installed at the same time?
> * from what i can tell it's persisting (plugin.save) package annotations in the ui-plugin metadata. this doesn't feel right. shouldn't the plugin's package information be available at other points?

Replying to your comments:
1) you are right. I don't know how I missed this. Will adjust accordingly.

2) First point: we cannot install two at the same time. The logic that copies the annotations are tied to a watcher, which will fire every time a plugin is installed, so it only runs once per install. Once it runs on this watcher, it's reset to `undefined`. It's again populated once an install occurs again. So, it should all be covered and work as expected.

3) It's the only way we can persist the data as it's the only resource available to us that is tied to plugins. Afaik, there's no other alternative and it's only metadata outside of the scope of the actual `uiplugins.metadata` field. This is something that is going to be useful to us to do more checks in the future.

richard-cox replied:
we need to run through this together. also spotted if the user navigates away from the page the extra information is never applied

aalves08 commented on shell/plugins/plugin.js:
```diff
@@ -29,41 +29,57 @@ export default async function(context) {
+        const shouldNotLoad = shouldNotLoadPlugin(plugin, rancherVersion, currKubeVersion, context.store.getters['uiplugins/plugins'] || []); // Error key string or boolean
+
+        if (!shouldNotLoad) {
+          hash[plugin.name] = context.$plugin.loadPluginAsync(plugin);
```
Will adjust code to contemplate that scenario. Thanks for the feedback.

aalves08 replied:
I believe these were never captured as the `try-catch` needed to be on the `allSettled` part (here we are just queueing up promises).

https://github.com/rancher/dashboard/blob/master/shell/plugins/plugin.js#L64

richard-cox replied:
there was no allSettled before, so the exception / failure would be caught by the try block

```
  try {
      const res = await context.store.dispatch('management/request', {
        url:                  `${ UI_PLUGIN_BASE_URL }/index.json`,
        method:               'GET',
        headers:              { accept: 'application/json' },
        redirectUnauthorized: false,
      });
```

aalves08 commented:
> I didn't get to test functional stuff because we might have some issues, specifically fetching the local cluster to get it's kube version
> 
> * user may not have permissions to see the local cluster
> * hopefully we should be able to load plugins on the log in screen in 2.8.0. at this point fetching the local cluster is impossible
> 
> Couple of options
> 
> * we ask that the local kube version be added to the `rancherversion` request (possible security issue?)
> * we ask the response to `http:ui-plugin-operator:80/proxy/index.json` to contain info on whether they're valid or not

Well, that's a good point @richard-cox . Ideal scenario, imho, would be to add it to the `rancherversion` request. It would make life so much easier.
The second option is not ideal, imo, as if we updated the local cluster kube version, we would need to ensure we updated all of the UIPlugin resources with that info. But it's definitely doable and probably quicker than waiting for something to change on the backend side. 

This can tracked in a separate issue, imo. What do you think @richard-cox ?

richard-cox commented on shell/assets/translations/en-us.yaml:
```diff
@@ -4172,6 +4173,8 @@ plugins:
     title: Add Extensions repositories
     prompt: You can install multiple Rancher extension repositories to increase your extensions catalog
+  kubeIncompatibility:
+    banner: Some installed extensions may not function property as they are incompatible with the Kubernetes version of your 'local' cluster
```
there was a typo in my suggestion `property` --> `properly`

richard-cox commented on shell/plugins/plugin.js:
```diff
@@ -29,41 +29,57 @@ export default async function(context) {
-          }
-        });
+        if (localCluster) {
+          currKubeVersion = localCluster.k3sConfig?.kubernetesVersion || '';
```
the advice to check model kubernetesVersion depends on the type of resource fetched. there'll be only one type, so no need for checking for different ones.

if the resource needs changing in to a model we should be able to classify it to access the kubernetesVersion properties

aalves08 replied:
classify `store.getters['management/classify'](obj)` and check for `kubernetesVersion` model property

aalves08 commented:
will be closed as we are dependent on backend and work done on this PR would fundamentally change. Even if it's closed, we still have this as record of the work done, which is a good basis for a new PR when backend get's the time to unblock UI)
